File: F#/chap07/Chap0706.fs
namespace Chap0706

type AsyncResult<'success, 'failure> = Async<Result<'success, 'failure>>

type UnvalidatedOrder = Undefined
type ValidatedOrder = Undefined
type ValidationError = Undefined
type PricedOrder = Undefined
type OrderAcknowledgmentSent = Undefined
type PricingError = Undefined
type PlaceOrderEvent = Undefined

type ValidateOrder =
    UnvalidatedOrder // In
        -> AsyncResult<ValidatedOrder, ValidationError> // Out

type PriceOrder =
    ValidatedOrder // In
        -> Result<PricedOrder, PricingError> // Out

type AcknowledgeOrder =
    PricedOrder // In
        -> Async<OrderAcknowledgmentSent option> // Out

type CreateEvents =
    PricedOrder // In
        -> PlaceOrderEvent list // Out

File: Scala/chap07/src/main/scala/Chap0706.scala
import cats.data.EitherT

import scala.concurrent.Future

object Chap0706:
  case class UnvalidatedOrder(value: Any)

  case class ValidatedOrder(value: Any)

  case class ValidationError(value: Any)

  case class PricedOrder(value: Any)

  case class OrderAcknowledgmentSent(value: Any)

  case class PricingError(value: Any)

  case class PlaceOrderEvent(value: Any)

  type ValidateOrder =
    UnvalidatedOrder // In
      => EitherT[Future, ValidationError, ValidatedOrder] // Out

  type PriceOrder =
    ValidatedOrder // In
      => Either[PricingError, PricedOrder] // Out

  type AcknowledgeOrder =
    PricedOrder // In
      => Future[Option[OrderAcknowledgmentSent]] // Out

  type CreateEvents =
    PricedOrder // In
      => List[PlaceOrderEvent] // Out
